tests/Stub/HtmlOptionsForm.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Form\Tests\Stub;
+
+use Yiisoft\Form\FormModel;
+use Yiisoft\Form\HtmlOptions\EmailHtmlOptions;
+use Yiisoft\Form\HtmlOptions\HasLengthHtmlOptions;
+use Yiisoft\Form\HtmlOptions\MatchRegularExpressionHtmlOptions;
+use Yiisoft\Form\HtmlOptions\NumberHtmlOptions;
+use Yiisoft\Form\HtmlOptions\RequiredHtmlOptions;
+use Yiisoft\Form\Tests\ValidatorFactoryMock;
+use Yiisoft\Validator\Rule\Email;
+use Yiisoft\Validator\Rule\HasLength;
+use Yiisoft\Validator\Rule\MatchRegularExpression;
+use Yiisoft\Validator\Rule\Number;
+use Yiisoft\Validator\Rule\Required;
+
+final class HtmlOptionsForm extends FormModel
+{
+    private string $number = '';
+    private string $hasLength = '';
+    private string $required = '';
+    private string $pattern = '';
+    private string $email = '';
+    private string $combined = '';
+
+    public function __construct()
+    {
+        parent::__construct(new ValidatorFactoryMock());
+    }
+
+    public function rules(): array
+    {
+        return [
+            'number' => [
+                $this->getNumberHtmlOptions(),
+            ],
+            'hasLength' => [
+                $this->getHasLengthHtmlOptions(),
+            ],
+            'required' => [
+                $this->getRequiredHtmlOptions(),
+            ],
+            'pattern' => [
+                $this->getMatchRegularExpressionHtmlOptions(),
+            ],
+            'email' => [
+                $this->getEmailHtmlOptions(),
+            ],
+            'combined' => [
+                $this->getNumberHtmlOptions(),
+                $this->getHasLengthHtmlOptions(),
+                $this->getRequiredHtmlOptions(),
+                $this->getMatchRegularExpressionHtmlOptions(),
+            ],
+        ];
+    }
+
+    private function getMatchRegularExpressionHtmlOptions(): MatchRegularExpressionHtmlOptions
+    {
+        return new MatchRegularExpressionHtmlOptions(
```

Comment:
How would that work for validation?

Reply:
OK, I see it in https://github.com/yiisoft/form/pull/68.

Is it possible not to require me to wrap all my rules into `HtmlOptionsProvider`? That's too repetitive and we don't need config options there.

Reply:
This is how the decorator pattern works. Merging this functionality into validation rules would be the wrong decision.
The rule settings are used in `HtmlOptionsProvider`'s, so the solution with the decorator was chosen for a reason.

If you would like to suggest another solution, I am ready to consider it.